pkg/artifactory/repositories.go

```diff
@@ -330,426 +332,533 @@ var repoTypesLikeGeneric = []string{
-	},
-	"default_deployment_repo": {
-		Type:        schema.TypeString,
-		Optional:    true,
-		Description: "Default repository to deploy artifacts.",
-	},
-	"retrieval_cache_period_seconds": {
-		Type:         schema.TypeInt,
-		Optional:     true,
-		Default:      7200,
-		Description:  "This value refers to the number of seconds to cache metadata files before checking for newer versions on aggregated repositories. A value of 0 indicates no caching.",
-		ValidateFunc: validation.IntAtLeast(0),
-	},
+		"artifactory_requests_can_retrieve_remote_artifacts": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: "Whether the virtual repository should search through remote repositories when trying to resolve an artifact requested by another Artifactory instance.",
+		},
+		"default_deployment_repo": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Default repository to deploy artifacts.",
+		},
+		"retrieval_cache_period_seconds": {
+			Type:         schema.TypeInt,
+			Optional:     true,
+			Default:      7200,
+			Description:  "This value refers to the number of seconds to cache metadata files before checking for newer versions on aggregated repositories. A value of 0 indicates no caching.",
+			ValidateFunc: validation.IntAtLeast(0),
+		},
+	}
+	return baseVirtualRepoSchema
+}
+
+func getBaseFederatedRepoSchema(packageType string) map[string]*schema.Schema {
+
+	var baseFederatedRepoSchema = map[string]*schema.Schema{
+		"key": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: repoKeyValidator,
+		},
+		"project_key": {
+			Type:             schema.TypeString,
+			Optional:         true,
+			ValidateDiagFunc: projectKeyValidator,
+			Description:      "Project key for assigning this repository to. When assigning repository to a project, repository key must be prefixed with project key, separated by a dash.",
+		},
+		"project_environments": {
+			Type:        schema.TypeSet,
+			Elem:        &schema.Schema{Type: schema.TypeString},
+			MinItems:    1,
+			MaxItems:    2,
+			Set:         schema.HashString,
+			Optional:    true,
+			Description: `Project environment for assigning this repository to. Allow values: "DEV" or "PROD"`,
+		},
+		"package_type": {
+			Type:     schema.TypeString,
+			Required: false,
+			Computed: true,
+			ForceNew: true,
+		},
+		"description": {
+			Type:     schema.TypeString,
+			Optional: true,
+		},
+		"notes": {
+			Type:     schema.TypeString,
+			Optional: true,
+		},
+		"includes_pattern": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"excludes_pattern": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+		},
+		"repo_layout_ref": {
+			Type:     schema.TypeString,
+			Optional: true,
+			DefaultFunc: func() (interface{}, error) {
```

Comment:
So, in your 
```go
 getDefaultFederatedRepoLayoutRef(packageType)
```

If you change it to:
you can return a function and get rid of this ugliness:
```go

var repoRefDefault = func() func() (interface{}, error) {
	return func() (interface{}, error) { 
                getDefaultFederatedRepoLayoutRef(packageType), nil
          }
}()
```
 then you can just do `DefaultFunc: repoRefDefault` -

Reply:
All the following functions have been removed to make the code clean.
```
getDefaultLocalRepoLayoutRef
getDefaultFederatedRepoLayoutRef
getDefaultVirtualRepoLayoutRef
getDefaultRemoteRepoLayoutRef
```
`getDefaultRepoLayoutRef` function has been implemented instead.

Reply:
please implement that function

Reply:
made the improvement.